sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -523,6 +529,17 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      // Null value should be handled by NullPropagation.
```

Comment:
It's better to make optimizer rules orthogonal and don't rely on each other. Let's make `isAlwaysFalse` stricter:
```
exps.forall  { e =>
  val res = EqualTo(_, equalTo).eval(EmptyRow)
  res != null && !res.asInstanceOf[Boolean]
}
```

Reply:
How about:
```scala
exps.forall {
  case l: Literal =>
    val res = EqualTo(l, equalTo).eval(EmptyRow)
    res != null && !res.asInstanceOf[Boolean]
  case _ => false
}
```

otherwise this test will fail:
```scala
val nonFoldable = CaseWhen(Seq(normalBranch, (a, b)), None)
assertEquivalent(EqualTo(nonFoldable, Literal(1)), EqualTo(nonFoldable, Literal(1)))
```